extensions/micrometer/runtime/src/main/java/io/quarkus/micrometer/runtime/binder/HttpCommonTags.java

```diff
@@ -46,30 +48,47 @@ public static Tag outcome(int statusCode) {
 
     /**
      * Creates a {@code uri} tag based on the URI of the given {@code request}.
-     * Falling back to {@code REDIRECTION} for 3xx responses, {@code NOT_FOUND}
-     * for 404 responses, {@code root} for requests with no path info, and {@code UNKNOWN}
+     * Falling back to {@code REDIRECTION} for 3xx responses if there wasn't a matched path pattern, {@code NOT_FOUND}
+     * for 404 responses if there wasn't a matched path pattern, {@code root} for requests with no path info, and
+     * {@code UNKNOWN}
      * for all other requests.
      *
      * @param pathInfo request path
+     * @param initialPath initial path before request pattern matching took place. Pass in null if there is pattern matching
+     *        done in the caller.
      * @param code status code of the response
      * @return the uri tag derived from the request
      */
-    public static Tag uri(String pathInfo, int code) {
-        if (code > 0) {
-            if (code / 100 == 3) {
-                return URI_REDIRECTION;
-            } else if (code == 404) {
-                return URI_NOT_FOUND;
-            }
-        }
+    public static Tag uri(String pathInfo, String initialPath, int code) {
         if (pathInfo == null) {
             return URI_UNKNOWN;
         }
         if (pathInfo.isEmpty() || "/".equals(pathInfo)) {
             return URI_ROOT;
         }
 
+        if (code > 0) {
+            if (code / 100 == 3) {
+                if (isTemplatedPath(pathInfo, initialPath)) {
+                    return Tag.of("uri", pathInfo);
+                } else {
+                    return URI_REDIRECTION;
+                }
+            } else if (code == 404) {
+                if (isTemplatedPath(pathInfo, initialPath)) {
+                    return Tag.of("uri", pathInfo);
+                } else {
+                    return URI_NOT_FOUND;
+                }
+            }
+        }
+
         // Use first segment of request path
         return Tag.of("uri", pathInfo);
     }
+
+    private static boolean isTemplatedPath(String pathInfo, String initialPath) {
+        // only include the path info if it has been matched to a template (initialPath != pathInfo) to avoid a metrics explosion with lots of entries
+        return initialPath != null && !Objects.equals(initialPath, pathInfo);
```

Comment:
I really wonder if this is desirable behavior @ahus1 @ebullient . When I call path `/test/` than it compares `/test` with `/test/` and decides that _it's template path_ and makes URI dynamic. IMO you should work with normalized paths, because same will go for empty path segments.